### PR TITLE
feat: implement audio warp/time-stretch feature

### LINKED ISSUE #958: feat: Add audio warp/time-stretch independent of pitch
## Problem

The SamplerEngine uses `playbackRate` for pitch shifting, which also changes the duration (higher pitch = shorter). Real DAWs can time-stretch audio independently of pitch — essential for tempo-matching loops, vocal tuning, and sample manipulation. Ableton's warp engine is a core differentiator.

## Current State

- `src/engine/SamplerEngine.ts` — Uses playbackRate (repitch mode only)
- `src/types/project.ts` — `StretchMode` type exists ('repitch' | 'slice') but 'slice' is not implemented
- No pitch-independent time stretching

## Acceptance Criteria

- [ ] Time-stretch mode: change tempo/duration without affecting pitch
- [ ] Pitch-shift mode: change pitch without affecting duration
- [ ] Warp markers: user-placeable markers to align beats to grid
- [ ] Auto-warp: detect BPM and place warp markers automatically
- [ ] Stretch algorithms: at minimum one decent quality algorithm
- [ ] Clips automatically stretch to project BPM when warp is enabled
- [ ] Visual: waveform with warp markers, draggable alignment

## Competitive Reference

- **Ableton Live**: Warp engine with Complex/Complex Pro/Beats/Tones/Texture modes
- **Logic Pro**: Flex Time with Polyphonic/Rhythmic/Monophonic/Speed modes
- **FL Studio**: NewTone + ZGameEditor pitch/time

## Technical Notes

- Phase vocoder is the standard algorithm for pitch-independent time stretch
- Libraries: `soundtouch-js` (LGPL, phase vocoder) or `rubberband.js` (GPL, high quality)
- AudioWorklet-based implementation for real-time performance
- Simpler approach: OLA (Overlap-Add) for rhythmic content, phase vocoder for tonal
- `OfflineAudioContext` can pre-render stretched audio for non-real-time use

### PR DESCRIPTION
**Pull request by ChuxiJ**
## Summary

- Add pitch shift algorithm (phase vocoder + resample) that preserves audio duration
- Integrate pitch shift into AudioEngine's stretched buffer pipeline
- Create WarpStretchSubmenu for clip context menu with stretch mode selection, pitch shift controls, BPM auto-detection, and project tempo matching
- Enhance ClipWarpMarkers with drag-to-reposition and alt+click-to-add interactivity
- Pass pitchShift through the transport scheduling pipeline

## Checklist (issue #958)

- [x] Independent tempo adjustment without pitch changes (5 time-stretch modes)
- [x] Pitch modification preserving duration (new `pitchShift()` algorithm)
- [x] User-placeable warp markers for beat alignment (drag + alt+click to add)
- [x] Automatic BPM detection and marker placement (`detectBpm` + UI trigger)
- [x] Quality stretch algorithms (beats, tones, complex, complexPro, texture)
- [x] Automatic clip stretching to project tempo ("Match Project Tempo" button)
- [x] Visual waveform interface with draggable markers (enhanced ClipWarpMarkers)

## Files changed (11 files, +744 lines)

**New:**
- `src/components/timeline/WarpStretchSubmenu.tsx` — Context menu submenu
- `src/components/timeline/__tests__/clipWarpMarkers.test.tsx` — 7 tests
- `src/components/timeline/__tests__/warpStretchSubmenu.test.tsx` — 3 tests

**Modified:**
- `src/utils/timeStretch.ts` — `pitchShift()` and `pitchShiftStereo()`
- `src/engine/AudioEngine.ts` — Pitch shift in stretched buffer pipeline
- `src/hooks/useTransport.ts` — Pass `pitchShift` through scheduling
- `src/components/timeline/ClipContextMenu.tsx` — Added WarpStretchSubmenu
- `src/components/timeline/ClipContextMenuContainer.tsx` — Pass clip prop
- `src/components/timeline/ClipWarpMarkers.tsx` — Drag + add interactivity
- `src/components/timeline/ClipBlock.tsx` — Enable `allowAdd` on markers
- `tests/unit/utils/timeStretch.test.ts` — 8 new pitch shift tests

## Test plan

- [x] 8 new pitchShift unit tests (duration preservation, energy, edge cases)
- [x] 2 new pitchShiftStereo tests
- [x] 3 new WarpStretchSubmenu render tests
- [x] 7 new ClipWarpMarkers interaction tests (drag, add, remove)
- [x] All 6054 existing tests pass (546 test files)
- [x] `npx tsc --noEmit` — 0 errors
- [x] `vite build` — succeeds

Closes #958

https://claude.ai/code/session_01Fj7L3DZSe2bfSaHXXF1hpU

### CODE REVIEW

File: src/components/timeline/ClipWarpMarkers.tsx [hunk 1]
```diff
@@ -16,29 +18,105 @@ const MARKER_WIDTH = 8;
...
+    const onMouseUp = () => {
+      dragStartRef.current = null;
+      setDraggingIndex(null);
+      document.removeEventListener('mousemove', onMouseMove);
+      document.removeEventListener('mouseup', onMouseUp);
+    };
+
+    document.addEventListener('mousemove', onMouseMove);
+    document.addEventListener('mouseup', onMouseUp);
+  }, [clipId, clipDuration, width, markers, setWarpMarkers]);
```
[the remark this answers is not in the document]
Comment (ChuxiJ) [reply src/components/timeline/ClipWarpMarkers.tsx]: Fixed in 51c80930. Added `dragCleanupRef` with `useEffect` cleanup that removes active document listeners on unmount.

---
_Generated by [Claude Code](https://claude.ai/code)_
